source/module_elecstate/module_charge/charge_init.cpp

```diff
@@ -53,61 +53,37 @@ void Charge::init_rho(elecstate::efermi& eferm_iout,
             {
                 std::stringstream ssc;
                 ssc << PARAM.globalv.global_readin_dir << "SPIN" << is + 1 << "_CHG.cube";
-                double& ef_tmp = eferm_iout.get_ef(is);
-                if (ModuleIO::read_cube(
-#ifdef __MPI
-                    & (GlobalC::Pgrid),
-#endif
+                if (ModuleIO::read_cube(GlobalC::Pgrid,
                     (PARAM.inp.esolver_type == "sdft" ? GlobalV::RANK_IN_STOGROUP : GlobalV::MY_RANK),
-                    is,
                     GlobalV::ofs_running,
-                    PARAM.inp.nspin,
                     ssc.str(),
                     this->rho[is],
-                    this->rhopw->nx,
-                    this->rhopw->ny,
-                    this->rhopw->nz,
-                    ef_tmp,
-                    & (GlobalC::ucell),
-                    this->prenspin))
+                    GlobalC::ucell.nat))
                 {
```

Comment:
I am still a littble bit confused about the content read from the cube file and seems there are some information dropped. The cube file contains lines of human readable comments, then followed by number of voxels in three direction, then as far as I remember, the atoms (their charges, positions?), then a three-folded for loop, from x to y then z.
So my question is, why I cannot see:
- human readable comments
- number of voxels and grids in three directions
- atoms' position and charges
?
The function read_cube should return what the file really contains, without any further selection/picking/heavy postprocessing on data, otherwise it is not a function that can be called read_cube, and the real/appropriate name cannot be easily determined because the function actually does more than one things.
Though it is good to see there are fewer parameters that are not directly relevant with the cube file.

Reply:
- The 2-line comments are skipped.
- The number of voxels are read into `nx_read, ny_read_nz_read` as before, and the following 3 numbers are skipped as before.
- Atom's position and charges are skipped as before.

If we need to read these two lines, where do you think to store them? (and note that users do not expect an error-quit when these info different from those in the workspace, so neither equality checks are needed)

Reply:
@maki49 for implementing functions, they should have their individual behaviors, what a function should do is to do all things that defined by its name. So if something is not needed presently, what is correct is to read it and save it to a temporary variable, then it will be unregistered/released after leaving the caller of the read_cube.

Reply:
OK, I will change `ignore` into `ifs >> temp`.

Reply:
@maki49 hmmm, not in this way, an appropriate example would be:
https://github.com/deepmodeling/abacus-develop/blob/fe3b44fe8c9b04195632ea844e32dc9cc19c5ae4/source/module_psi/psi_initializer_nao.cpp#L88